### PR TITLE
Scheduler

### PR DESCRIPTION
**Pull request by engzongwei**
## Context

Updated function that republishes task when scheduler supervisor notifies that previous task has failed.

## Test
Can be tested with @Jaylin0312 Supervisor PR. 
1. start supervisor MS & scheduler
2. publish message to task.status
3. check that new task has been added into task.execute queue

### CODE REVIEW

File: scheduler/scheduler.py [hunk 1]
```diff
@@ -30,11 +30,17 @@ def checkType(msg):
     elif msg['type'] == 'milestone.reward' or msg['type'] == 'milestone.penalise':
         # Milestone has been rewarded, need to remove from cron
         milestone_id = msg['resource_id']
-        projId = msg['data']['project_id']
+        projId = msg['data']['id']
         milestoneRemove(milestone_id, projId)
-    elif msg['resource_id'] == 'task.add':
+    elif msg['type'] in ["upcoming", "penalise","rollback"]:
```
Comment (zhiweit) [inline scheduler/scheduler.py:35]: if type is 'penalise' i.e. the milestone 1 day after its due_date and verifier havent verify, you should publish task with event 'overdue'. Currently it publishes a payload of 'rollback' to the projectpolice.
To test: publish this to `scheduler.task.add` queue
```json
{
  "resource_id": "a3cc8e03-07e4-4211-a029-431beed4ef87",
  "type": "penalise",
  "data": {
    "project_id": "a3cc8e03-07e4-4211-a029-431beed4ef87",
    "milestone_id": "92636c2d-d979-4351-b75a-865e3dd3d5b6",
    "payment_id": null
  }
}

payload published to `task_execute`
```json
{
  "resource_id": "a3cc8e03-07e4-4211-a029-431beed4ef87",
  "type": "rollback",
  "data": {
    "project_id": "a3cc8e03-07e4-4211-a029-431beed4ef87",
    "milestone_id": "92636c2d-d979-4351-b75a-865e3dd3d5b6",
    "payment_id": null
  }
}
```

File: scheduler/publisher.py [hunk 2]
```diff
@@ -36,15 +36,17 @@ def check_setup(connection, channel, RMQHOSTNAME, RMQPORT):
...
-        "milestone_id": milestone_id
+        "milestone_id": milestone_id,
+        "payment_id": payment_id
     }
     return json.dumps({"resource_id": resource_id, "type": type, "data": new_data})
 
 # Function that republishes tasks that failed
-def publishTask(event, project_id, milestone_id):
+def publishTask(event, project_id, milestone_id, payment_intent_id=None):
     type = ""
```
Comment (zhiweit) [inline scheduler/publisher.py:50]: Why change the type here? Why not just pass the event as the type?

**Review by zhiweit — CHANGES_REQUESTED** [review]
Tested the queues by publishing to the queue and the cron publishes tasks to `task_execute` queue accordingly. 
You may refer to the bottom of Project MS notion on what payload I passed to test the queue to test this PR.

1. Include an instruction to change the end-of-line (EOL) to 'LF' from 'CRLF' for Windows users when testing it for the `start.sh` file
2. The consumer stops consuming from the queues after 5 mins of inactivity (not consuming messages). Change it so that it will continue to consume messages indefinitely.


For subsequent PRs, please include how to test (sample testing data and what output to expect) for reviewers to test the PR 😅.

File: scheduler/scheduler.py (hunk 1 above)
Comment (engzongwei) [reply scheduler/scheduler.py:35]: image
image
image

It should already be event = "overdue"

File: scheduler/scheduler.py [hunk 3]
```diff
@@ -49,8 +55,8 @@ def removeOffsetTrack(payment_id):
 def newOffsetTrack(payment_id, created_at, project_id, milestone_id):
     print('in scheduler newOffsetTrack')
     cron = CronTab(user=True)
-    job  = cron.new(command=f'/usr/local/bin/python /app/publisher.py --type offset --proj {project_id} --mile {milestone_id}', comment=f'offset_{payment_id}')
-    job.setall(datetime.fromisoformat(created_at) + timedelta(hours=1))
+    job  = cron.new(command=f'/usr/local/bin/python /app/publisher.py --type offset --proj {project_id} --mile {milestone_id} --payment {payment_id}', comment=f'offset_{payment_id}')
+    job.setall(datetime.strptime(created_at,"%Y-%m-%dT%H:%M:%SZ") + timedelta(hours=1))
     # Can be used for testing, will schedule the job to run in 1 minute
     # job.setall(datetime.fromisoformat(due_date) + timedelta(minutes=1))
```
[the remark this answers is not in the document]
Comment (engzongwei) [reply scheduler/scheduler.py]: Thanks for the suggested improvements. Will update accordingly

File: scheduler/scheduler.py [hunk 4]
```diff
@@ -49,8 +55,8 @@ def removeOffsetTrack(payment_id):
 def newOffsetTrack(payment_id, created_at, project_id, milestone_id):
     print('in scheduler newOffsetTrack')
     cron = CronTab(user=True)
-    job  = cron.new(command=f'/usr/local/bin/python /app/publisher.py --type offset --proj {project_id} --mile {milestone_id}', comment=f'offset_{payment_id}')
-    job.setall(datetime.fromisoformat(created_at) + timedelta(hours=1))
+    job  = cron.new(command=f'/usr/local/bin/python /app/publisher.py --type offset --proj {project_id} --mile {milestone_id} --payment {payment_id}', comment=f'offset_{payment_id}')
+    job.setall(datetime.strptime(created_at,"%Y-%m-%dT%H:%M:%SZ") + timedelta(hours=1))
```
[the remark this answers is not in the document]
Comment (engzongwei) [reply scheduler/scheduler.py]: done. thanks

File: scheduler/scheduler.py (hunk 1 above)
Comment (engzongwei) [reply scheduler/scheduler.py:38]: Good spot. I have edited the code to reflect this.

File: scheduler/publisher.py (hunk 2 above)
Comment (engzongwei) [reply scheduler/publisher.py:50]: I have updated the code to simplify the logic. do check and let me know if there should be anymore changes

**Comment by engzongwei** [comment]
Closing this PR. Refactored Scheduler merged into main

**Pull request closed without merging**